File: BOJ/minji/10_week/타겟넘버.py
from collections import deque

def solution(numbers, target):
    
    def dfs(idx, current_sum):
        
        if idx == len(numbers):# idx가 numbers의 길이에 도달하면(모든 숫자를 다 처리했으면)
            return 1 if current_sum == target else 0 # 현재까지의 합 current_sum이 target과 같으면 1을 반환, 아니면 0을 반환
        
        #현재 숫자를 더하는 경우 - dfs(idx + 1, current_sum + numbers[idx])
        #현재 숫자를 빼는 경우 - dfs(idx + 1, current_sum - numbers[idx])
        return dfs(idx + 1, current_sum + numbers[idx]) + dfs(idx + 1, current_sum - numbers[idx])

    return dfs(0, 0)

File: BOJ/jihoon/10week/타겟 넘버.cpp
﻿/*** DFS 풀이 ***/

#include <bits/stdc++.h>

using namespace std;

int answer = 0;
void dfs(vector<int> numbers, int target, int idx, int sum) {
    if (idx == numbers.size()) { // 끝에 도달
        if (sum == target) { // target 완성
            answer++;
        }
        return;
    }

    dfs(numbers, target, idx + 1, sum + numbers[idx]); // +1
    dfs(numbers, target, idx + 1, sum - numbers[idx]); // -1
}

int solution(vector<int> numbers, int target) {

    dfs(numbers, target, 0, 0);

    return answer;
}


/*** BFS 풀이 ***/

/*
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> numbers, int target){
    int answer = 0;
    queue<int> Q;
    Q.push(numbers[0]);
    Q.push(-1 * numbers[0]);

    for(int i=1; i<numbers.size(); i++){
        int len = Q.size();
        for(int j=0; j<len; j++){
            int sum = Q.front();
            Q.pop();
            Q.push(sum + numbers[i]);
            Q.push(sum - numbers[i]);
        }
    }
    int len = Q.size();
    for(int i=0; i<len; i++){
        if(Q.front() == target)
            answer++;
        Q.pop();
    }

    return answer;
}
*/